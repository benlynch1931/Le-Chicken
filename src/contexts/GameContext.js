import React, { createContext, Component } from 'react';
import { widthPercentageToDP as wp, heightPercentageToDP as hp } from 'react-native-responsive-screen';

export const GameContext = createContext();

const chickenPositionDefault = [wp("43.34%"), hp("60%")];
class GameContextProvider extends Component {
  state = {
    currentScene: 'coop',
    chickenGraphic: 'idleright',
    chickenPosition: chickenPositionDefault,
    inputText: "",
    hint: "Pour marcher: Type ‘marcher’",
    chickenToMove: 0,
    level: 0,
    chickenDirection: "up",
    walls: [{
      type: 'horizontal',
      position: 48.34,
      start: 56.67,
      end: 100.00,
      stroke: 1.85
    }, {
      type: 'vertical',
      position: 2.00,
      start: 0.00,
      end: 49.26,
      stroke: 4.00
    }]
  }



  changeLevel = (level) => {
    this.setState({ level: level })
  }
  changeChickenDirection = (direction) => {
    this.setState({ chickenDirection: direction })
  }
  changeInputText = (inputText) => {
    this.setState({ inputText: inputText })
  }
  changeScene = (sceneName) => {
    this.setState({ currentScene: sceneName })
  }
  changeChickenGraphic = (chickenGraphic) => {
    this.setState({ chickenGraphic: chickenGraphic })
  }
  increaseChickenPosition = (x, y) => {
    this.setState({ chickenPosition: [this.state.chickenPosition[0] + x, this.state.chickenPosition[1] + y] })
  }
  resetChickenPosition = () => {
    this.setState({ chickenPosition: chickenPositionDefault })
    this.changeChickenGraphic('up')
  }

  changeHint = (hint) => {
    this.setState({ hint: hint })
  }

  changeChickenToMove = (chickenToMove) => {
    this.setState({ chickenToMove: chickenToMove })
  }

  render() {
    return (
      <GameContext.Provider value={{
        ...this.state,
        changeScene: this.changeScene,
        changeChickenGraphic: this.changeChickenGraphic,
        increaseChickenPosition: this.increaseChickenPosition,
        resetChickenPosition: this.resetChickenPosition,
        changeInputText: this.changeInputText,
        changeHint: this.changeHint,
        changeChickenToMove: this.changeChickenToMove,
        changeChickenDirection: this.changeChickenDirection,
        changeLevel: this.changeLevel
      }}>
        {this.props.children}
      </GameContext.Provider>
    );
  }
}

export default GameContextProvider;
