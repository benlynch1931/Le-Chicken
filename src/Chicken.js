import React, { useContext, useEffect } from 'react';
import { Image } from 'react-native';
import { GameContext } from './contexts/GameContext.js';
import { widthPercentageToDP as wp, heightPercentageToDP as hp } from 'react-native-responsive-screen';


const Chicken = () => {
  const { chickenPosition, dPadPressed, level, chickenGraphic, changeChickenGraphic, increaseChickenPosition, chickenDirection, changeChickenToMove, chickenToMove, walls, currentScene, needToUpdateChickenGraphic, changeNeedToUpdateChickenGraphic } = useContext(GameContext)
  const chickenWidth = wp("11.73%")
  const chickenHeight = hp("5.42%")
  const stepSizeVertical = hp("0.5%")
  const stepSizeHorizontal = wp("1.0%")



  const _moveIncrement = (direction, distance) => {

    switch (direction) {
      case 'up':
        increaseChickenPosition(0, -stepSizeVertical)
        break;
      case 'down':
        increaseChickenPosition(0, stepSizeVertical)
        break;
      case 'right':
        increaseChickenPosition(stepSizeHorizontal, 0)
        break;
      case 'left':
        increaseChickenPosition(-stepSizeHorizontal, 0)
        break;
    }
  }

  const _finishMovement = (direction, chickenWalk) => {
    clearInterval(chickenWalk)
    changeChickenToMove(0);
    triggerGraphicChange()
  }

  const triggerGraphicChange = () => {
    if (!dPadPressed) {
      changeNeedToUpdateChickenGraphic(true);
    }
  }

  const stepSize = (direction) => {
    if (isVertical(direction)) {
      return stepSizeVertical
    } else {
      return stepSizeHorizontal
    }
  }

  const isVertical = (direction) => {
    return (direction == 'up' || direction == 'down')
  }

  const capDistanceForWall = (wall, distance, direction) => {
    const padding = linePadding(wall.stroke, direction)
    const wallPosition = adjustCoords(wall.position, direction, padding)
    if (chickenWillReach(wallPosition, distance, direction) && chickenInLineWith(wall)) {
      distance = Math.floor((Math.abs(chickenEdge(direction) - wallPosition)) / stepSize(direction))
    }
    return distance
  }

  const move = (direction, distance) => {
    if (currentScene == 'maze') {
      walls.forEach(wall => distance = capDistanceForWall(wall, distance, direction))
    }

    let chickenWalk = setInterval(() => {
      if (distance <= 0) {
        _finishMovement(direction, chickenWalk);
        return;
      }
      _moveIncrement(direction);
      distance--;
    }, 30)
  }

  const linePadding = (stroke, direction) => {
    switch (direction) {
      case 'up':
        return stroke / 2
      case 'down':
        return -stroke / 2
      case 'right':
        return -stroke / 2
      case 'left':
        return stroke / 2
    }
  }

  const chickenWillReach = (wallPosition, distance, direction) => {
    let chickenStartsBeforeWall
    let chickenEndsBeyondWall
    switch (direction) {
      case 'up':
        chickenStartsBeforeWall = chickenEdge(direction) >= wallPosition - stepSizeVertical
        chickenEndsBeyondWall = chickenPositionAfterMovement(distance, direction) <= wallPosition + stepSizeVertical
        break;
      case 'down':
        chickenStartsBeforeWall = chickenEdge(direction) <= wallPosition + stepSizeVertical
        chickenEndsBeyondWall = chickenPositionAfterMovement(distance, direction) >= wallPosition - stepSizeVertical
        break;
      case 'right':
        chickenStartsBeforeWall = chickenEdge(direction) <= wallPosition + stepSizeHorizontal
        chickenEndsBeyondWall = chickenPositionAfterMovement(distance, direction) >= wallPosition - stepSizeHorizontal
        break;
      case 'left':
        chickenStartsBeforeWall = chickenEdge(direction) >= wallPosition - stepSizeHorizontal
        chickenEndsBeyondWall = chickenPositionAfterMovement(distance, direction) <= wallPosition + stepSizeHorizontal
        break;
    }
    return chickenStartsBeforeWall && chickenEndsBeyondWall
  }

  const chickenEdge = (direction) => {
    switch (direction) {
      case 'up':
        return chickenPosition[1]
      case 'down':
        return chickenPosition[1] + chickenHeight
      case 'right':
        return chickenPosition[0] + chickenWidth
      case 'left':
        return chickenPosition[0]
    }
  }

  const chickenPositionAfterMovement = (distance, direction) => {
    switch (direction) {
      case 'up':
        return chickenEdge(direction) - (distance * stepSizeVertical)
      case 'down':
        return chickenEdge(direction) + (distance * stepSizeVertical)
      case 'right':
        return chickenEdge(direction) + (distance * stepSizeHorizontal)
      case 'left':
        return chickenEdge(direction) - (distance * stepSizeHorizontal)
    }
  }

  const chickenInLineWith = (wall) => {
    switch (wall.type) {
      case 'horizontal':
        return chickenPosition[0] > adjustXCoords(wall.start) && chickenPosition[0] < adjustXCoords(wall.end)
      case 'vertical':
        // is the chicken more than the start, and less than the end of the line?
        return chickenPosition[1] > adjustYCoords(wall.start) && chickenPosition[1] < adjustYCoords(wall.end)
    }
  }

  const adjustCoords = (coord, direction, padding = 0) => {
    if (direction == 'up' || direction == 'down') {
      return adjustYCoords(coord, padding)
    } else {
      return adjustXCoords(coord, padding)
    }
  }

  const adjustXCoords = (coord, padding = 0) => {
    return wp(coord) + wp(padding)
  }

  const adjustYCoords = (coord, padding = 0) => {
    return hp(coord) + hp('5.00%') + hp("1.85%") + hp(padding)
  }

  useEffect(() => {
    if (chickenToMove == 0) {
      return;
    }
    move(chickenDirection, chickenToMove);
  }, [level, chickenToMove, chickenDirection])

  useEffect(() => {
    let action
    if (chickenToMove > 0) {
      action = 'walk'
    } else {
      action = 'idle'
    }
    if (needToUpdateChickenGraphic == true) {
      changeChickenGraphic(`${action}${chickenDirection}`)
      changeNeedToUpdateChickenGraphic(false)
    }
  }, [chickenToMove, chickenDirection, needToUpdateChickenGraphic])


  const graphics = {
    idleleft: require('../assets/chicken-left.png'),
    idleright: require('../assets/chicken-right.png'),
    idleup: require('../assets/chicken-stand-back.png'),
    idledown: require('../assets/chicken-stand-front.png'),
    walkup: require('../assets/chicken-run-back.gif'),
    walkright: require('../assets/chicken-run-right.gif'),
    walkleft: require('../assets/chicken-run-left.gif'),
    walkdown: require('../assets/chicken-run-front.gif'),
  }


  return (
    < Image
      style={{
        position: 'absolute',
        top: chickenPosition[1],
        left: chickenPosition[0],
        width: chickenWidth,
        height: chickenHeight,
        zIndex: 4,
      }}
      nativeID={`chicken-${chickenGraphic}`}
      source={graphics[chickenGraphic]}
    />
  );

}

export default Chicken
