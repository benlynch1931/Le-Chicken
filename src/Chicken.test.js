import React from 'react';
import { create, act } from 'react-test-renderer';
import Chicken from './Chicken.js';
import { GameContext } from './contexts/GameContext.js';

const mockContext = { 
    chickenPositionY: 0,
    level: 0,
    chickenGraphic: "walkUp",
    changeChickenGraphic: jest.fn(),
    increaseChickenPositionY: jest.fn(),
    chickenMoving: false,
    changeChickenMoving: jest.fn()
}

test('renders correctly when the level is 0', () => {
    const chicken = create(<GameContext.Provider value={mockContext}><Chicken /></GameContext.Provider>)
    expect(mockContext.increaseChickenPositionY).not.toBeCalled();
    expect(mockContext.changeChickenGraphic).not.toBeCalled();
    expect(chicken).toMatchSnapshot();
  });

test('renders correctly when the level is 1', async () => {
  mockContext.level = 1
  jest.useFakeTimers();
  await act(async () => {
    create(<GameContext.Provider value={mockContext}><Chicken /></GameContext.Provider>)
  });
  expect(mockContext.changeChickenGraphic).toHaveBeenCalled();
  jest.advanceTimersByTime(3000);
  expect(mockContext.increaseChickenPositionY).toHaveBeenCalled();
});
