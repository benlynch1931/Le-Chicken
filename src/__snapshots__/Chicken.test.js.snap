// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly when the level is 0 1`] = `
<Image
  nativeID="chicken-walkUp"
  style={
    Object {
      "height": 82,
      "left": 0,
      "position": "absolute",
      "top": 0,
      "width": 100,
      "zIndex": 4,
    }
  }
/>
`;

exports[`renders correctly when the level is 1 1`] = `
<Image
  nativeID="chicken-walkUp"
  style={
    Object {
      "height": 82,
      "left": 0,
      "position": "absolute",
      "top": 0,
      "width": 100,
      "zIndex": 4,
    }
  }
/>
`;
